src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java

```diff
@@ -0,0 +1,272 @@
+		input1.add( SolutionMappingUtils.createSolutionMapping(
+				var1, x2,
+				var2, y2) );
+
+		final GenericIntermediateResultBlockImpl input2 = new GenericIntermediateResultBlockImpl();
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var2, y1,
+				var3, z1) );
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var2, y1,
+				var3, z2) );
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var2, y2,
+				var3, z3) );
+
+		final Iterator<SolutionMapping> it = runTest(input1, input2);
+
+		// checking
+		final Set<Binding> result = new HashSet<>();
+
+		assertTrue( it.hasNext() );
+		result.add( it.next().asJenaBinding() );
+
+		assertTrue( it.hasNext() );
+		result.add( it.next().asJenaBinding() );
+
+		assertTrue( it.hasNext() );
+		result.add( it.next().asJenaBinding() );
+
+		assertFalse( it.hasNext() );
+
+		boolean b1Found = false;
+		boolean b2Found = false;
+		boolean b3Found = false;
+		for ( final Binding b : result ) {
+			assertEquals( 3, b.size() );
+
+			if ( b.get(var1).getURI().equals("http://example.org/x1") ) {
+				assertEquals( "http://example.org/y1", b.get(var2).getURI() );
+				if ( b.get(var3).getURI().equals("http://example.org/z1") ) {
+					b1Found = true;
+				}
+				else if ( b.get(var3).getURI().equals("http://example.org/z2") ) {
+					b2Found = true;
+				}
+				else {
+					fail( "Unexpected URI for ?v3: " + b.get(var3).getURI() );
+				}
+			}
+			else if ( b.get(var1).getURI().equals("http://example.org/x2") ) {
+				assertEquals( "http://example.org/y2", b.get(var2).getURI() );
+				assertEquals( "http://example.org/z3", b.get(var3).getURI() );
+				b3Found = true;
+			}
+			else {
+				fail( "Unexpected URI for ?v1: " + b.get(var1).getURI() );
+			}
+		}
+
+		assertTrue(b1Found);
+		assertTrue(b2Found);
+		assertTrue(b3Found);
+
+	}
+
+	protected void _joinWithTwoJoinVariables() {
+		final Var var1 = Var.alloc("v1");
+		final Var var2 = Var.alloc("v2");
+		final Var var3 = Var.alloc("v3");
+
+		final Node x1 = NodeFactory.createURI("http://example.org/x1");
+		final Node x2 = NodeFactory.createURI("http://example.org/x2");
+		final Node y1 = NodeFactory.createURI("http://example.org/y1");
+		final Node y2 = NodeFactory.createURI("http://example.org/y2");
+		final Node z1 = NodeFactory.createURI("http://example.org/z1");
+		final Node z2 = NodeFactory.createURI("http://example.org/z2");
+
+		final GenericIntermediateResultBlockImpl input1 = new GenericIntermediateResultBlockImpl();
+		input1.add( SolutionMappingUtils.createSolutionMapping(
+				var1, x1,
+				var2, y1) );
+		input1.add( SolutionMappingUtils.createSolutionMapping(
+				var1, x2,
+				var2, y2) );
+
+		final GenericIntermediateResultBlockImpl input2 = new GenericIntermediateResultBlockImpl();
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var1, x1,
+				var2, y1,
+				var3, z1) );
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var1, x2,
+				var2, y2,
+				var3, z2) );
```

Comment:
Please add another unit test where the join is on two variables and input solution mappings have the same value for one of the variables but not for the other (i.e., there are no join partners).

Reply:
Thanks for adding that other unit test!